### PR TITLE
Add DSA Validations

### PR DESCRIPTION
**Pull request by AntoxaAntoxic**
https://github.com/prebid/prebid-server/issues/3438

### CODE REVIEW

File: src/main/java/org/prebid/server/auction/DsaEnforcer.java [hunk 1]
```diff
@@ -71,9 +97,58 @@ private static boolean isDsaValidationRequired(BidRequest bidRequest) {
...
+        }
+
+        validateFieldLength(dsaResponse);
+
+        final Integer adRender = dsaResponse.getAdRender();
+        final Integer pubRender = Optional.ofNullable(bidRequest.getRegs())
+                .map(Regs::getExt)
+                .map(ExtRegs::getDsa)
+                .map(ExtRegsDsa::getPubRender)
+                .orElse(null);
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/auction/DsaEnforcer.java:112]:
```
final Integer pubRender = bidRequest.getRegs().getExt().getDsa().getPubRender();
```

File: src/main/java/org/prebid/server/auction/DsaEnforcer.java [hunk 2]
```diff
@@ -71,9 +97,58 @@ private static boolean isDsaValidationRequired(BidRequest bidRequest) {
...
+
+        if (pubRender.equals(DsaPublisherRender.WILL_RENDER.getValue())
+                && adRender != null && adRender.equals(DsaAdvertiserRender.WILL_RENDER.getValue())) {
+            throw new PreBidException("DSA publisher and buyer both signal will render");
+        }
+
+        if (pubRender.equals(DsaPublisherRender.NOT_RENDER.getValue())
+                && (adRender == null || adRender.equals(DsaAdvertiserRender.NOT_RENDER.getValue()))) {
+            throw new PreBidException("DSA publisher and buyer both signal will not render");
+        }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/auction/DsaEnforcer.java:126]:
```
        if (pubRender == DsaPublisherRender.WILL_RENDER.getValue()
                && adRender != null && adRender == DsaAdvertiserRender.WILL_RENDER.getValue()) {
            throw new PreBidException("DSA publisher and buyer both signal will render");
        }

        if (pubRender == DsaPublisherRender.NOT_RENDER.getValue()
                && (adRender == null || adRender != DsaAdvertiserRender.WILL_RENDER.getValue())) {
            throw new PreBidException("DSA publisher and buyer both signal will not render");
        }
```

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/DsaPublisherRender.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    private final Integer value;
+
+    DsaPublisherRender(final Integer value) {
+        this.value = value;
+    }
+
+    public Integer getValue() {
+        return value;
+    }
+}
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/DsaPublisherRender.java:18]:
```
public enum DsaPublisherRender {

    NOT_RENDER(0),
    COULD_RENDER(1),
    WILL_RENDER(2);

    private final int value;

    DsaPublisherRender(int value) {
        this.value = value;
    }

    public int getValue() {
        return value;
    }
}
```

File: src/main/java/org/prebid/server/proto/openrtb/ext/response/DsaAdvertiserRender.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    private final Integer value;
+
+    DsaAdvertiserRender(final Integer value) {
+        this.value = value;
+    }
+
+    public Integer getValue() {
+        return value;
+    }
+}
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/proto/openrtb/ext/response/DsaAdvertiserRender.java:17]:
```
public enum DsaAdvertiserRender {

    NOT_RENDER(0),
    WILL_RENDER(1);

    private final int value;

    DsaAdvertiserRender(int value) {
        this.value = value;
    }

    public int getValue() {
        return value;
    }
}

```

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/DsaRequired.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    private final Integer value;
+
+    DsaRequired(final Integer value) {
+        this.value = value;
+    }
+
+    public Integer getValue() {
+        return value;
+    }
+}
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/DsaRequired.java:19]:
```
public enum DsaRequired {

    NOT_REQUIRED(0),
    SUPPORTED(1),
    REQUIRED(2),
    REQUIRED_ONLINE_PLATFORM(3);

    private final int value;

    DsaRequired(int value) {
        this.value = value;
    }

    public int getValue() {
        return value;
    }
}
```

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/auction/DsaEnforcer.java (hunk 2 above)
Comment (AntoxaAntoxic) [reply src/main/java/org/prebid/server/auction/DsaEnforcer.java:126]: never mind

File: src/main/java/org/prebid/server/auction/DsaEnforcer.java [hunk 6]
```diff
@@ -24,7 +30,16 @@
...
-    private static final Set<Integer> DSA_REQUIRED = Set.of(2, 3);
+    private static final Set<Integer> DSA_REQUIRED = Set.of(
+            DsaRequired.REQUIRED.getValue(),
+            DsaRequired.REQUIRED_ONLINE_PLATFORM.getValue());
+    private static final int MAX_DSA_FIELD_LENGTH = 100;
+
+    private final JacksonMapper mapper;
+
+    public DsaEnforcer(JacksonMapper mapper) {
+        this.mapper = mapper;
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/auction/DsaEnforcer.java:41]: `Objects.requireNotNull()`

File: src/main/java/org/prebid/server/auction/DsaEnforcer.java [hunk 7]
```diff
@@ -71,9 +96,54 @@ private static boolean isDsaValidationRequired(BidRequest bidRequest) {
...
 
-    private boolean isValid(Bid bid) {
-        final ObjectNode bidExt = bid.getExt();
-        return bidExt != null && bidExt.hasNonNull(DSA_EXT) && !bidExt.get(DSA_EXT).isEmpty();
+    private static void validateDsa(BidRequest bidRequest, ExtBidDsa dsaResponse) {
+        if (dsaResponse == null) {
+            throw new PreBidException("DSA object missing when required");
+        }
+
+        validateFieldLength(dsaResponse);
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/auction/DsaEnforcer.java:104]: not needed anymore, we already checked this earlier

**Review by CTMBNara — CHANGES_REQUESTED** [review]

**Review by CTMBNara — APPROVED** [review]

**Pull request merged**